test/widgets/docs_screenshots/conftest.py

```diff
@@ -0,0 +1,128 @@
+    docs_folder = (
+        Path(__file__).parent
+        / ".."
+        / ".."
+        / ".."
+        / "docs"
+        / "_static"
+        / "screenshots"
+        / "widgets"
+    )
+    log = os.path.join(docs_folder, "shots.json")
+    if os.path.isdir(folder):
+        shutil.rmtree(folder)
+    os.makedirs(folder)
+    key = {}
+
+    def get_file_name(w_name, config):
+        nonlocal key
+
+        # Convert config into a string of key=value
+        entry = ", ".join(f"{k}={repr(v)}" for k, v in config.items())
+
+        # Check if widget is in the key dict
+        if w_name not in key:
+            key[w_name] = {}
+
+        # Increment the index number
+        indexes = [int(x) for x in key[w_name]]
+        index = max(indexes) + 1 if indexes else 1
+
+        # Record the config
+        key[w_name][index] = entry
+
+        # Define the target folder and check it exists
+        shots_dir = os.path.join(folder, w_name)
+        if not os.path.isdir(shots_dir):
+            os.mkdir(shots_dir)
+
+        # Returnt the path for the screenshot
+        return os.path.join(shots_dir, f"{index}.png")
+
+    yield get_file_name
+
+    # We copy the screenshots from the test folder to the docs folder at the end
+    # This prevents pytest deleting the files itself
+
+    # Remove old screenshots
+    if os.path.isdir(docs_folder):
+        shutil.rmtree(docs_folder)
+
+    # Copy to the docs folder
+    shutil.copytree(folder, docs_folder)
+    with open(log, "w") as f:
+        json.dump(key, f)
+
+    # Clear up the tests folder
+    shutil.rmtree(folder)
+
+
+@pytest.fixture
+def screenshot_manager(widget, request, manager_nospawn, minimal_conf_noscreen, target):
+    """
+    Create a manager instance for the screenshots. Individual "tests" should only call
+    `screenshot_manager.take_screenshot()` but the destination path is also available in
+    `screenshot_manager.target`.
+
+    Widgets should create their own `widget` fixture in the relevant file (applying
+    monkeypatching etc as necessary).
+
+    Configs can then be passed by parametrizing "screenshot_manager".
+    """
+    # Get the widget and config
+    config = getattr(request, "param", dict())
+    wdgt = widget(**config)
+    name = wdgt.name
+
+    # Create a function to generate filename
+    def filename():
+        return target(name, config)
+
+    # Add the widget to our config
+    minimal_conf_noscreen.groups = [Group(i) for i in "123456789"]
+    minimal_conf_noscreen.fake_screens = [
+        Screen(top=Bar([wdgt], 32), x=0, y=0, width=300, height=300),
+        Screen(top=Bar([], 32), x=0, y=300, width=300, height=300),
+    ]
+
+    manager_nospawn.start(minimal_conf_noscreen)
+
+    # Add some convenience attributes for taking screenshots
+    manager_nospawn.target = filename
+    ss_widget = manager_nospawn.c.widget[name]
+    manager_nospawn.take_screenshot = lambda f=filename: ss_widget.take_screenshot(f())
```

Comment:
re the `cmd_take_screenshot` commands above, they could be moved into this file and added to `minimal_conf_noscreen`.

Also - as it's all being run with X11 only, you can probably remove the `qtile.core.name` check from those functions and just hard-code x11 in there.

Reply:
Thinking about this a bit. We need the `take_screenshot` commands to be exposed via IPC. When we move the decorators, commands can't be added to existing objects as the parsing is done at creation. 

However, I think we can still make this work by creating new classes in the fixture which inherit from `Bar` and the relevant `widget`. Putting decorators on these should work fine.